subsys/net/lib/lwm2m/lwm2m_obj_gateway.c

```diff
@@ -146,6 +147,57 @@ static struct lwm2m_engine_obj_inst *lwm2m_gw_create(uint16_t obj_inst_id)
 	return &inst[index];
 }
 
+int lwm2m_gw_handle_req(struct lwm2m_message *msg) 
+{
+	struct lwm2m_engine_obj_inst *obj_inst;
+	struct coap_option options[4];
+	int ret;
+
+	ret = coap_find_options(msg->in.in_cpkt, COAP_OPTION_URI_PATH, options, ARRAY_SIZE(options));
+	if (ret < 0) {
+		return ret;
+	}
+
+	for (int obj_inst_id = 0; obj_inst_id == CONFIG_LWM2M_GATEWAY_MAX_INSTANCES - 1;
+	     obj_inst_id++) {
+		void *prefix = NULL;
+		uint16_t prefix_len = 0;
+
+		obj_inst = get_engine_obj_inst(LWM2M_OBJECT_GATEWAY_ID, obj_inst_id);
+		if (obj_inst == NULL) {
+			return -ENOENT;
+		}
+		lwm2m_get_res_buf(
+			&LWM2M_OBJ(LWM2M_OBJECT_GATEWAY_ID, obj_inst_id, LWM2M_GATEWAY_PREFIX_RID),
+			&prefix, NULL, &prefix_len, NULL);
+
+		if (strncmp(options[0].value, (char *)prefix, prefix_len) != 0 || prefix == NULL) {
+			return -ENOENT;
+		}
+
+		if (gateway_msg_cb[obj_inst_id] == NULL) {
+			return -ENOENT;
+		}
+
+		/* Delete prefix from path*/
+		ret = coap_options_to_path(&options[1], ret - 1, &msg->path);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = gateway_msg_cb[obj_inst_id](msg);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+	return 0;
```

Comment:
After a second thought, do we need to return here one of the 3 options that might have happened:
1. GW object did not handle the request. This is now signalled as negative code.
2. Callback handled the request, but did not return response.
3. Callback handled the requested and returned a response in `msg`.

Now only option 1 is handled, but I'm unsure if we assume 2 or 3 option when we return 0.
Please check `

Reply:
i've thought about it again and i think it makes sense to leave the transmission of the msg to the lwm2m engine. 
The engine is the "gateway" and therefore, in my opinion, it should also take over the transmission. The cb handles the complete coap msg and its formatting. This also allows us to reuse the existing coap responses for certain error codes.